applications/asset_tracker/src/main.c

```diff
@@ -1437,7 +1437,9 @@ void main(void)
 			continue;
 		}
 
-		if (ret == 0) {
+		if ((ret == 0) &&
+		    (cloud_keepalive_time_left(cloud_backend) == 0)) {
```

Comment:
Do you know if the bug that makes `poll()` return even though timeout hasn't passed has been reported somewhere?
Note that there's a check wihtin `mqtt_live()` that checks if a ping should be sent, so even though we call `cloud_ping()`, it's not a given that a ping is actually sent. 
`mqtt_live()` now returns `-EAGAIN` if a ping was not sent, but unfortunately it's not propagated through nRF Cloud lib and it's glue layer to the cloud API. If it was, we could have checked the return value to know if ping was sent.

Reply:
I didn't report the poll() bug yet, or research very hard if it had been.  I will take a look now.

Yes, I noticed the not-returning-the-error thing.  Agreed.  The trouble is the time elapsed_getting reset when it shouldn't.

Reply:
I dug further into the poll() returning early issue, and turns out my initial analysis missed an important thing.  Turns out, poll() is fine.  What's going on is that poll() is called with a timeout value, and then in another thread, data is sent to the cloud some time later, but well before poll() returns.  That means that the ping could be delayed further, but it won't be without my change, because the original call to poll() is still active.

But, my change does help, because what it does it makes sure that the ping() occurs the keepalive time after the most recent LTE transfer, so it will save bandwidth when the keepalive is faster than other data transfers.  It also prevents unnecessary pings when other data is output at a faster rate than the keepalive (though presumably, the underlying mqtt_ping() prevents this already).

I'll post some logging in the Jira ticket.

Reply:
I updated the git commit message to reflect this.

Reply:
Why is this change needed, because this is already handled by the MQTT stack (mqtt_live() in mqtt.c)? It only sends a ping when keepalive time has passed since last data send. There shouldn't be any unnecessary pings.

Reply:
@tokangas remember that this is calling cloud_ping(), not mqtt_live() directly.  The design goal of the cloud api is to make it backend agnostic.  So I think it needs to be here even though yes, mqtt_live() does appear to fix it.

Besides, in my opinion, designs that rely on deep knowledge of hidden behaviors of multiple layers under them are brittle and prone to breaking when those other layers change.

I think it is better when it is more obvious and clear to the reader of the code what is happening, without having to learn a vast amount of other people's code under it just to make sense of it.

Reply:
I see your point on cloud backend agnositicity (is that an actual word?) and not needing to know the deep secrets of the ping function.